plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/expression/SparkExpressionBuilder.java

```diff
@@ -0,0 +1,196 @@
+
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+
+public class SparkExpressionBuilder
+        extends SparkExpressionBaseVisitor<Object>
+{
+    public SparkExpressionBuilder() {}
+
+    @Override
+    public Object visitStandaloneExpression(SparkExpressionParser.StandaloneExpressionContext context)
+    {
+        return visit(context.expression());
+    }
+
+    @Override
+    public Object visitExpression(SparkExpressionParser.ExpressionContext context)
+    {
+        return super.visitExpression(context);
+    }
+
+    @Override
+    public Object visitPredicated(SparkExpressionParser.PredicatedContext context)
+    {
+        if (context.predicate() != null) {
+            return visit(context.predicate());
+        }
+
+        return visit(context.valueExpression);
+    }
+
+    @Override
+    public Object visitComparison(SparkExpressionParser.ComparisonContext context)
+    {
+        return new ComparisonExpression(
+                getComparisonOperator(((TerminalNode) context.comparisonOperator().getChild(0)).getSymbol()),
+                (Expression) visit(context.value),
+                (Expression) visit(context.right));
+    }
+
+    @Override
+    public Node visitAnd(SparkExpressionParser.AndContext context)
+    {
+        List<ParserRuleContext> terms = flatten(context, element -> {
+            if (element instanceof SparkExpressionParser.AndContext) {
+                SparkExpressionParser.AndContext and = (SparkExpressionParser.AndContext) element;
+                return Optional.of(and.booleanExpression());
+            }
+
+            return Optional.empty();
+        });
+
+        return new LogicalExpression(LogicalExpression.Operator.AND, visit(terms, Expression.class));
+    }
+
+    private static List<ParserRuleContext> flatten(ParserRuleContext root, Function<ParserRuleContext, Optional<List<? extends ParserRuleContext>>> extractChildren)
+    {
+        List<ParserRuleContext> result = new ArrayList<>();
+        Deque<ParserRuleContext> pending = new ArrayDeque<>();
+        pending.push(root);
+
+        while (!pending.isEmpty()) {
+            ParserRuleContext next = pending.pop();
+
+            Optional<List<? extends ParserRuleContext>> children = extractChildren.apply(next);
+            if (!children.isPresent()) {
+                result.add(next);
+            }
+            else {
+                for (int i = children.get().size() - 1; i >= 0; i--) {
+                    pending.push(children.get().get(i));
+                }
+            }
+        }
+
+        return result;
+    }
```

Comment:
That looks complicated. Maybe we just support binary AND and OR boolean expression?
I.e. always two operands, not a list?

In Trino expressions, there is a list used for performance reasons only.
We can be simpler here, it doesn't matter.